File: java_apx/src/test/java/net/sympower/citizen/apx/DataRequestServiceTest.java
package net.sympower.citizen.apx;

import com.github.tomakehurst.wiremock.client.WireMock;
import de.mkammerer.wiremock.WireMockExtension;
import net.sympower.citizen.apx.dto.incomming.Quote;
import net.sympower.citizen.apx.dto.incomming.QuoteResponse;
import net.sympower.citizen.apx.service.DataRequestService;
import org.apache.commons.io.FileUtils;
import org.eclipse.jetty.http.HttpStatus;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.RegisterExtension;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.core.io.Resource;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.web.reactive.function.client.WebClientResponseException;

import java.io.IOException;

import static com.github.tomakehurst.wiremock.client.WireMock.equalTo;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;



@SpringBootTest
class DataRequestServiceTest {

    @RegisterExtension
    static final WireMockExtension wireMock = new WireMockExtension();

    @Autowired
    private DataRequestService service;

    @Value("classpath:apx-data.json")
    private Resource apxData;

    @Test
    void shouldGetResponseForMockedProvider() throws IOException {
        // GIVEN
        String jsonString = FileUtils.readFileToString(apxData.getFile());
        String mockUrl = getPreconfiguredMockWebServerUrl("/some/endpoint", HttpStatus.OK_200, jsonString);
        // WHEN
        QuoteResponse response = service.getProviderData(mockUrl);
        // THEN
        assertThat(response.getQuote()).hasSize(24);
        Quote quote = response.getQuote().get(4);
        assertThat(quote.getMarket()).isEqualTo("APX Power NL Hourly");
        assertThat(quote.getDate_applied()).isEqualTo(1573599600000L);
        assertThat(quote.getValues()).hasSize(4);
        assertThat(quote.getValues().get(0).gettLabel()).isEqualTo("Order");
        assertThat(quote.getValues().get(0).getValue()).isEqualTo("5");
        assertThat(quote.getValues().get(1).gettLabel()).isEqualTo("Hour");
        assertThat(quote.getValues().get(1).getValue()).isEqualTo("05");
        assertThat(quote.getValues().get(2).gettLabel()).isEqualTo("Net Volume");
        assertThat(quote.getValues().get(2).getValue()).isEqualTo("4838.70");
        assertThat(quote.getValues().get(3).gettLabel()).isEqualTo("Price");
        assertThat(quote.getValues().get(3).getValue()).isEqualTo("34.72");
        quote = response.getQuote().get(18);
        assertThat(quote.getMarket()).isEqualTo("APX Power NL Hourly");
        assertThat(quote.getDate_applied()).isEqualTo(1573599600000L);
        assertThat(quote.getValues()).hasSize(4);
        assertThat(quote.getValues().get(0).gettLabel()).isEqualTo("Order");
        assertThat(quote.getValues().get(0).getValue()).isEqualTo("19");
        assertThat(quote.getValues().get(1).gettLabel()).isEqualTo("Hour");
        assertThat(quote.getValues().get(1).getValue()).isEqualTo("19");
        assertThat(quote.getValues().get(2).gettLabel()).isEqualTo("Net Volume");
        assertThat(quote.getValues().get(2).getValue()).isEqualTo("3384.30");
        assertThat(quote.getValues().get(3).gettLabel()).isEqualTo("Price");
        assertThat(quote.getValues().get(3).getValue()).isEqualTo("56.09");
    }

    @Test
    void shouldThrowExceptionByProviderIssues() {
        // GIVEN
        String mockUrl = getPreconfiguredMockWebServerUrl("/broken/endpoint", HttpStatus.BAD_GATEWAY_502, "{ \"error\": \"BOOOOOM!\" }");
        // WHEN
        WebClientResponseException exception = assertThrows(WebClientResponseException.class, () -> service.getProviderData(mockUrl));
        // THEN
        assertThat(String.format("502 Bad Gateway from GET %s", mockUrl)).isEqualTo(exception.getMessage());
    }


    private String getPreconfiguredMockWebServerUrl(String endpointUrl, int status, String body) {
        wireMock.stubFor(
                WireMock
                        .get(endpointUrl)
                        .withHeader(HttpHeaders.CONTENT_TYPE, equalTo(MediaType.APPLICATION_JSON_VALUE))
                        .willReturn(
                                WireMock.aResponse()
                                        .withHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE)
                                        .withStatus(status)
                                        .withBody(body)
                        )
        );
        return wireMock
                .getBaseUri()
                .resolve(endpointUrl)
                .toString();
    }

}

File: src/test/kotlin/net/sympower/citizen/apx/DataRequestServiceTest.kt
package net.sympower.citizen.apx

import com.github.tomakehurst.wiremock.client.WireMock
import com.github.tomakehurst.wiremock.client.WireMock.equalTo
import de.mkammerer.wiremock.WireMockExtension
import net.sympower.citizen.apx.service.DataRequestService
import org.apache.commons.io.FileUtils
import org.assertj.core.api.Assertions.assertThat
import org.eclipse.jetty.http.HttpStatus
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.RegisterExtension
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.core.io.Resource
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType
import org.springframework.web.reactive.function.client.WebClientResponseException


@SpringBootTest
class DataRequestServiceTest {

    @JvmField
    @RegisterExtension
    final var wireMock = WireMockExtension()

    @Autowired
    private lateinit var service: DataRequestService

    @Value("classpath:apx-data.json")
    private lateinit var apxData: Resource


    @Test
    fun `should get response from mocked provider`() {
        // GIVEN
        val jsonString = FileUtils.readFileToString(apxData.file)
        val mockUrl = getPreconfiguredMockWebServer("/some/endpoint", HttpStatus.OK_200, jsonString)
        // WHEN
        val response = service.getProviderData(mockUrl)
        // THEN
        assertThat(response?.quote).hasSize(24)
        assertThat( response?.quote?.get(4)).isNotNull
        response?.quote?.get(4)?.let {
            assertThat(it.market).isEqualTo("APX Power NL Hourly")
            assertThat(it.date_applied).isEqualTo(1573599600000)
            assertThat(it.values).hasSize(4)
            assertThat(it.values[0].tLabel).isEqualTo("Order")
            assertThat(it.values[0].value).isEqualTo("5")
            assertThat(it.values[1].tLabel).isEqualTo("Hour")
            assertThat(it.values[1].value).isEqualTo("05")
            assertThat(it.values[2].tLabel).isEqualTo("Net Volume")
            assertThat(it.values[2].value).isEqualTo("4838.70")
            assertThat(it.values[3].tLabel).isEqualTo("Price")
            assertThat(it.values[3].value).isEqualTo("34.72")
        }
        assertThat( response?.quote?.get(4)).isNotNull
        response?.quote?.get(18)?.let {
            assertThat(it.market).isEqualTo("APX Power NL Hourly")
            assertThat(it.date_applied).isEqualTo(1573599600000)
            assertThat(it.values).hasSize(4)
            assertThat(it.values[0].tLabel).isEqualTo("Order")
            assertThat(it.values[0].value).isEqualTo("19")
            assertThat(it.values[1].tLabel).isEqualTo("Hour")
            assertThat(it.values[1].value).isEqualTo("19")
            assertThat(it.values[2].tLabel).isEqualTo("Net Volume")
            assertThat(it.values[2].value).isEqualTo("3384.30")
            assertThat(it.values[3].tLabel).isEqualTo("Price")
            assertThat(it.values[3].value).isEqualTo("56.09")
        }
    }

    @Test
    fun `should throw exception by provider issues`() {
        // GIVEN
        val mockUrl = getPreconfiguredMockWebServer("/broken/endpoint", HttpStatus.BAD_GATEWAY_502, "{ \"error\": \"BOOOOOM!\" }")
        // WHEN
        val exception = assertThrows(WebClientResponseException::class.java) { service.getProviderData(mockUrl) }
        // THEN
        assertThat("502 Bad Gateway from GET $mockUrl").isEqualTo(exception.message)
    }

    private fun getPreconfiguredMockWebServer(endpointUrl: String, status: Int, body: String): String {
        wireMock.stubFor(
            WireMock
                .get(endpointUrl)
                .withHeader(HttpHeaders.CONTENT_TYPE, equalTo(MediaType.APPLICATION_JSON_VALUE))
                .willReturn(
                    WireMock.aResponse()
                        .withHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE)
                        .withStatus(status)
                        .withBody(body)
                )
        )
        return wireMock
            .baseUri
            .resolve(endpointUrl)
            .toString()
    }

}